pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -0,0 +1,1087 @@
+	///   and transition in the post transition phase. To be safe and to not
+	///   make any unhandled assumptions, we throw `DispatchError::Other` for
+	///   these states though we need to make sure this can never occur!
+	fn handle_increase(
+		&self,
+		swap: Swap<T::Balance, T::CurrencyId>,
+	) -> Result<Self, DispatchError> {
+		if swap.currency_in == swap.currency_out {
+			return Self::handle_increase_non_foreign(self, swap);
+		}
+
+		match &self {
+			Self::NoState => Ok(Self::ActiveSwapIntoPoolCurrency { swap }),
+			// Add pool swap
+			Self::InvestmentOngoing { invest_amount } => {
+				Ok(Self::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+					swap,
+					invest_amount: *invest_amount,
+				})
+			}
+			// Bump pool swap
+			Self::ActiveSwapIntoPoolCurrency { swap: pool_swap } => {
+				swap.ensure_currencies_match(pool_swap, true)?;
+				Ok(Self::ActiveSwapIntoPoolCurrency {
+					swap: Swap {
+						amount: swap.amount.ensure_add(pool_swap.amount)?,
+						..swap
+					},
+				})
+			}
+			// Reduce foreign swap amount by the increasing amount and increase investing amount as
+			// well adding foreign_done amount by the minimum of active swap amounts
+			Self::ActiveSwapIntoForeignCurrency { swap: foreign_swap } => {
+				swap.ensure_currencies_match(foreign_swap, false)?;
+				let foreign_amount_pool_denominated = T::CurrencyConverter::stable_to_stable(
+					swap.currency_in,
+					swap.currency_out,
+					foreign_swap.amount,
+				)?;
+				let pool_amount_foreign_denominated = T::CurrencyConverter::stable_to_stable(
+					swap.currency_out,
+					swap.currency_in,
+					swap.amount,
+				)?;
+
+				match swap.amount.cmp(&foreign_amount_pool_denominated) {
+					// Pool swap amount is immediately fulfilled, i.e. invested and marked as done into foreign
+					Ordering::Less => {
+						Ok(Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+							swap: Swap {
+								amount: foreign_swap
+									.amount
+									.ensure_sub(pool_amount_foreign_denominated)?,
+								..*foreign_swap
+							},
+							done_amount: pool_amount_foreign_denominated,
+							invest_amount: swap.amount,
+						})
+					}
+					// Both opposite swaps are immediately fulfilled, i.e. invested and marked as done
+					Ordering::Equal => Ok(Self::SwapIntoForeignDoneAndInvestmentOngoing {
+						done_swap: *foreign_swap,
+						invest_amount: swap.amount,
+					}),
+					// Foreign swap amount is immediately fulfilled, i.e. invested and marked as done
+					Ordering::Greater => {
+						Ok(
+							Self::ActiveSwapIntoPoolCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+								swap: Swap {
+									// safe since amount_in_foreign > foreign_swap.amount
+									amount: swap.amount.ensure_sub(foreign_amount_pool_denominated)?,
+									..swap
+								},
+								done_amount: foreign_swap.amount,
+								invest_amount: swap.amount,
+							},
+						)
+					}
```

Comment:
I think, in this case, the `invest_amount` should be equal to `foreign_amount_pool_denominated`

My reasoning is that if you're in `ActiveSwapIntoForeignCurrency {foreign_swap}  and comes a `pool_swap` greater than the existent `foreign_swap`, you only can ensure that the possible invested amount after that action is `foreign_swap` you had previously, no more.

Reply:
Great catch, that is correct and probably a result of me rushing this. Thanks so much for all your findings and brainjuice.

Reply:
Thanks you for all the feedback and guide! It's to much rushing for getting this perfectly...